internal/sdkv2provider/data_source_jwks_from_certificate.go

```diff
@@ -0,0 +1,164 @@
+package sdkv2provider
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"encoding/pem"
+	"math/big"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+type Keys struct {
+	Keys []Key `json:"keys"`
+}
+
+type Key struct {
+	Kty    *string  `json:"kty"`
+	X5C    []string `json:"x5c"`
+	N      *string  `json:"n"`
+	E      *string  `json:"e"`
+	Kid    *string  `json:"kid"`
+	X5T256 *string  `json:"x5t#256"`
+	X5Dn   *string  `json:"x5dn"`
+}
+
+func dataSourceJwksFromCertificate() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceJwksFromCertificateRead,
+		Schema:      dataSourceJwksFromCertificateSchema(),
+		Description: `Calculates a JSON Web Key Set from a given certificate.`,
+	}
+}
+
+func dataSourceJwksFromCertificateSchema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		"pem": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: `Requires a pem encoded certificate.`,
+		},
+		"treat_independently": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     false,
+			Description: `Determines whether to generate JWK for only leaf certificate in a chain and put other certificates into x5c (default) or generate JWKs for each certificate in a chain`,
+		},
+		"jwks": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: `The calculated JWKS`,
+		},
+	}
+}
+
+func dataSourceJwksFromCertificateRead(_ context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+
+	pemString := d.Get("pem").(string)
+	treatIndependently := d.Get("treat_independently").(bool)
+
+	chain := decodePem(pemString)
+
+	var keys []Key
+
+	certificates := parseChain(chain.Certificate)
+
+	if treatIndependently {
```

Comment:
this logic should form part of the provider, im happy for chains to by default be correctly configured in the `x5c` but if people want to merge jwks keys, they should do that in the hcl.

Reply:
Left only default